# Análise de crédito

Rodando o primeiro dataset do livro __Projetos de ciêndia de dados com Python__

In [1]:
import pandas as pd
import numpy as np

### Exercício 2

Carregamento e leitura do tamanho do DataFrame

In [2]:
df = pd.read_excel('default_of_credit_card_clients__courseware_version_1_21_19.xls')
df

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE PAY_1  PAY_2  \
0      798fc410-45c1      20000    2          2         1   24     2      2   
1      8a8c8f3b-8eb4     120000    2          2         2   26    -1      2   
2      85698822-43f5      90000    2          2         2   34     0      0   
3      0737c11b-be42      50000    2          2         1   37     0      0   
4      3b7f77cc-dbc0      50000    1          2         1   57    -1      0   
...              ...        ...  ...        ...       ...  ...   ...    ...   
29995  ecff42d0-bdc6     220000    1          3         1   39     0      0   
29996  99d1fa0e-222b     150000    1          3         2   43    -1     -1   
29997  95cdd3e7-4f24      30000    1          2         2   37     4      3   
29998  00d03f02-04cd      80000    1          3         1   41     1     -1   
29999  15d69f9f-5ad3      50000    1          2         1   46     0      0   

       PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  \
0         -1     -1  ...          0          0          0         0       689   
1          0      0  ...       3272       3455       3261         0      1000   
2          0      0  ...      14331      14948      15549      1518      1500   
3          0      0  ...      28314      28959      29547      2000      2019   
4         -1      0  ...      20940      19146      19131      2000     36681   
...      ...    ...  ...        ...        ...        ...       ...       ...   
29995      0      0  ...      88004      31237      15980      8500     20000   
29996     -1     -1  ...       8979       5190          0      1837      3526   
29997      2     -1  ...      20878      20582      19357         0         0   
29998      0      0  ...      52774      11855      48944     85900      3409   
29999      0      0  ...      36535      32428      15313      2078      1800   

       PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
0             0         0         0         0                           1  
1          1000      1000         0      2000                           1  
2          1000      1000      1000      5000                           0  
3          1200      1100      1069      1000                           0  
4         10000      9000       689       679                           0  
...         ...       ...       ...       ...                         ...  
29995      5003      3047      5000      1000                           0  
29996      8998       129         0         0                           0  
29997     22000      4200      2000      3100                           1  
29998      1178      1926     52964      1804                           1  
29999      1430      1000      1000      1000                           1  

[30000 rows x 25 columns]

In [3]:
df.shape

(30000, 25)

### Exercício 3

Entendendo algumas características básicas do DataFrame: quais são as colunas, seus significados e a quantidade de IDs únicos

In [4]:
df.columns

Index(['ID', 'LIMIT_BAL', 'SEX', 'EDUCATION', 'MARRIAGE', 'AGE', 'PAY_1',
       'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6', 'BILL_AMT1', 'BILL_AMT2',
       'BILL_AMT3', 'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6', 'PAY_AMT1',
       'PAY_AMT2', 'PAY_AMT3', 'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6',
       'default payment next month'],
      dtype='object')

Entendendo os dados da coluna:

`ID`: Código identificador do cliente

`LIMIT_BAL`: Valor do crédito fornecido ao cliente

`SEX`: Gênero. 1 - Homem, 2 - Mulher 

`EDUCATION`: Nível educacional. 1 - Pós graduação, 2 - Ensino superior, 3 - Ensino médio, 4 - Outros

`MARRIAGE`: Estado civil. 1 - Casado, 2 - Solteiro, 3 - Outros

`AGE`: Idade 

`PAY_1 - PAY_6`: Registros de pagamentos de meses passados. A escala de medida dos pagamentos é:

-2: A conta começou o mês sem valor a ser pago<br>
-1: Pagamento sem atraso<br>
0: Pagamento mínimo foi feito. Saldo devedor não foi pago<br>
1: Atraso de um mês no pagamento<br>
...<br>
9: Atraso de nove meses ou mais no pagamento

`BILL_AMT1 - BILL_AMT6`: Valor da fatura no mês.

`PAY_AMT1 - PAY_AMT6`: Valor pago no mês.

In [5]:
df['ID'].nunique()

29687

#### Comentário 1

Aqui podemos ver que o número de ID's únicos é 29.687, abaixo dos 30.000 IDs da tabela inicial. Isso indica que alguns IDs estão duplicados ou repetidos mais vezes. Para saber a natureza deles e como proceder com esses dados, é necessário antes verificar quantos estão repetidos.

In [6]:
id_counts = df['ID'].value_counts()
id_counts

ad23fe5c-7b09    2
1fb3e3e6-a68d    2
89f8f447-fca8    2
7c9b7473-cc2f    2
90330d02-82d9    2
                ..
2edeb3a6-d38d    1
27e11893-92e9    1
2b1c282b-441e    1
bd737997-0eb0    1
15d69f9f-5ad3    1
Name: ID, Length: 29687, dtype: int64

In [7]:
id_counts.value_counts()

1    29374
2      313
Name: ID, dtype: int64

#### Comentário 2

Vemos aqui que 313 IDs se repetem duas vezes, 29374 só uma vez e nenhum se repete 3 vezes ou mais

### Máscaras booleanas

In [8]:
np.random.seed(seed = 24)
random_integers = np.random.randint(low = 1, high = 5, size = 100)
random_integers[:5]

array([3, 4, 1, 4, 2])

In [9]:
is_equal_to_3 = random_integers == 3
is_equal_to_3[:5]

array([ True, False, False, False, False])

In [10]:
sum(is_equal_to_3)

22

In [11]:
random_integers[is_equal_to_3]

array([3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3])

### Exercício 4

Buscando os IDs duplicados no DataFrame original. Para isso, criamos uma máscara booleana `dupe_mask` que receberá todos os IDs do DataFrame original e marcará como `True` aqueles que possuem 2 entradas no DataFrame original e como `False` todas as outras. Com essa máscara booleana iremos criar uma lista `dupe_ids` que contém todos os IDs duplicados.

In [12]:
dupe_mask = (id_counts == 2)
dupe_mask

ad23fe5c-7b09     True
1fb3e3e6-a68d     True
89f8f447-fca8     True
7c9b7473-cc2f     True
90330d02-82d9     True
                 ...  
2edeb3a6-d38d    False
27e11893-92e9    False
2b1c282b-441e    False
bd737997-0eb0    False
15d69f9f-5ad3    False
Name: ID, Length: 29687, dtype: bool

In [24]:
dupe_ids = list(id_counts.index[dupe_mask])
print(dupe_ids[:5])
print(len(dupe_ids))

['ad23fe5c-7b09', '1fb3e3e6-a68d', '89f8f447-fca8', '7c9b7473-cc2f', '90330d02-82d9']
313


Este comando é um pouco mais complexo. O objetivo dele é usar a lista dupe_ids para fazer uma busca no DataFrame original
para trazer as linhas cujo ID está duplicado. O comando df.loc realiza a busca baseada nos parâmetros passados.
Esses parâmetros, por sua vez, são `df['ID']`, uma vez que estamos usando o ID como ponto de filtragem, `.isin(duple_ids)` para buscar somente os elementos que fazem parte da lista `duple_ids`, e o `sort_values('ID')` pois queremos ordenar a nova tabela com base nos IDs para que IDs iguais fiquem juntos, facilitando assim a leitura.

Podemos ver que a tabela final possui 626 elementos. Esse número era esperado pois, como vimos acima, o DataFrame original possui 313 IDs duplicados.

In [14]:
df.loc[df['ID'].isin(dupe_ids)].sort_values('ID')

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE          PAY_1  \
15863  01355f46-3f0b          0    0          0         0    0              0   
15763  01355f46-3f0b     100000    2          2         2   23  Not available   
21972  0170086d-7472          0    0          0         0    0              0   
21872  0170086d-7472     100000    2          1         1   25              0   
3284   02e41927-a6df      60000    2          2         2   21              0   
...              ...        ...  ...        ...       ...  ...            ...   
20226  ff8b7170-ce3a          0    0          0         0    0              0   
2005   ffa0f012-3bd7      20000    1          2         2   34  Not available   
2105   ffa0f012-3bd7          0    0          0         0    0              0   
8746   ffb9a143-c375      80000    1          1         2   27              0   
8846   ffb9a143-c375          0    0          0         0    0              0   

       PAY_2  PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  \
15863      0      0      0  ...          0          0          0         0   
15763      2      2      2  ...      96332      97560      95885      5000   
21972      0      0      0  ...          0          0          0         0   
21872      0     -1      0  ...      52079      40183      35796        28   
3284       0      0      0  ...      25475      26010      26183      1859   
...      ...    ...    ...  ...        ...        ...        ...       ...   
20226      0      0      0  ...          0          0          0         0   
2005      -2     -2     -2  ...       2280       1500        780      1530   
2105       0      0      0  ...          0          0          0         0   
8746       0      0      0  ...      51179      52073      52833      5000   
8846       0      0      0  ...          0          0          0         0   

       PAY_AMT2  PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  \
15863         0         0         0         0         0   
15763      4300      3600      3700         0      7300   
21972         0         0         0         0         0   
21872    103350      1722      1100      1300      2000   
3284       1410       910       939      1000      1000   
...         ...       ...       ...       ...       ...   
20226         0         0         0         0         0   
2005        750      1140      3000       780         0   
2105          0         0         0         0         0   
8746       1801      5000      2035      2100      6200   
8846          0         0         0         0         0   

       default payment next month  
15863                           0  
15763                           0  
21972                           0  
21872                           0  
3284                            0  
...                           ...  
20226                           0  
2005                            0  
2105                            0  
8746                            0  
8846                            0  

[626 rows x 25 columns]

Vemos então que todas as contas duplicadas possuem uma linha com valores reais e uma com valores igual a 0 para todas as features. Essas segundas linhas podem (e devem) ser excluídas, uma vez que elas não possuem qualquer significado real.

In [25]:
df_zero_mask = (df == 0)
df_zero_mask

ID  LIMIT_BAL    SEX  EDUCATION  MARRIAGE    AGE  PAY_1  PAY_2  \
0      False      False  False      False     False  False  False  False   
1      False      False  False      False     False  False  False  False   
2      False      False  False      False     False  False   True   True   
3      False      False  False      False     False  False   True   True   
4      False      False  False      False     False  False  False   True   
...      ...        ...    ...        ...       ...    ...    ...    ...   
29995  False      False  False      False     False  False   True   True   
29996  False      False  False      False     False  False  False  False   
29997  False      False  False      False     False  False  False  False   
29998  False      False  False      False     False  False  False  False   
29999  False      False  False      False     False  False   True   True   

       PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  \
0      False  False  ...       True       True       True      True     False   
1       True   True  ...      False      False      False      True     False   
2       True   True  ...      False      False      False     False     False   
3       True   True  ...      False      False      False     False     False   
4      False   True  ...      False      False      False     False     False   
...      ...    ...  ...        ...        ...        ...       ...       ...   
29995   True   True  ...      False      False      False     False     False   
29996  False  False  ...      False      False       True     False     False   
29997  False  False  ...      False      False      False      True      True   
29998   True   True  ...      False      False      False     False     False   
29999   True   True  ...      False      False      False     False     False   

       PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
0          True      True      True      True                       False  
1         False     False      True     False                       False  
2         False     False     False     False                        True  
3         False     False     False     False                        True  
4         False     False     False     False                        True  
...         ...       ...       ...       ...                         ...  
29995     False     False     False     False                        True  
29996     False     False      True      True                        True  
29997     False     False     False     False                       False  
29998     False     False     False     False                       False  
29999     False     False     False     False                       False  

[30000 rows x 25 columns]

O DataFrame acima é uma máscara booleana onde todos os valores igual a 0 serão marcados como True. Isso não é bastante, uma vez que alguns valores realmente podem ser iguais a 0 mesmo para linhas válidas. Importante notar também que não haverá nenhuma linha na coluna ID marcada como True, uma vez que não é possível existis ID 0.

O próximo passo é filtrar, a partir dessa tabela, as linhas que possuem 0 para todas as colunas (exceto ID) e removê-las. Uma opção é criar a `feature_zero_mask`. O que acontece nessa feature é uma seleção `.iloc` do DataFrame `df_zero_mask`. Para o `.iloc` é necessário passar os parâmetros de seleção. Nesse caso temos `[:, 1:]`. Aqui estamos fazendo duas filtragrens. A primeira, marcada somente pelo `:`, delimita que todas as linhas serão analisadas. A segunda, `1:`, delimita que só serão analisadas as colunas a partir da indexada pelo índice 1 (ou seja, a segunda, `LIMIT_BAL`) até o final. Por fim, o método `.all(axis = 1)` nos mostra uma lista simples caso todas as colunas (`axis = 1`) sejam True, exceto a primeira, como já foi especificado anteriormente. Dessa forma, todas as linhas marcadas como True nesse último processo podem ser consideradas linhas duplicadas com valores igual a 0 para todas as features, exceto ID

In [16]:
feature_zero_mask = df_zero_mask.iloc[:, 1:].all(axis = 1)
feature_zero_mask

0        False
1        False
2        False
3        False
4        False
         ...  
29995    False
29996    False
29997    False
29998    False
29999    False
Length: 30000, dtype: bool

In [17]:
sum(feature_zero_mask)

315

In [18]:
df_clean = df.loc[~ feature_zero_mask, :].copy()

In [19]:
df_clean.shape[0] == df_clean['ID'].nunique()

True

Agora podemos ver que o tamanho do DataFrame `df_clean` é igual o número de IDs únicos. Ou seja, nosso DataFrame está limpo e organizado.

### Exercício 5

In [26]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 29685 entries, 0 to 29999
Data columns (total 25 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   ID                          29685 non-null  object
 1   LIMIT_BAL                   29685 non-null  int64 
 2   SEX                         29685 non-null  int64 
 3   EDUCATION                   29685 non-null  int64 
 4   MARRIAGE                    29685 non-null  int64 
 5   AGE                         29685 non-null  int64 
 6   PAY_1                       29685 non-null  object
 7   PAY_2                       29685 non-null  int64 
 8   PAY_3                       29685 non-null  int64 
 9   PAY_4                       29685 non-null  int64 
 10  PAY_5                       29685 non-null  int64 
 11  PAY_6                       29685 non-null  int64 
 12  BILL_AMT1                   29685 non-null  int64 
 13  BILL_AMT2                   29685 non-null  in

In [28]:
df_clean['PAY_1'].head(10)

0     2
1    -1
2     0
3     0
4    -1
5     0
6     0
7     0
8     0
9    -2
Name: PAY_1, dtype: object

In [29]:
df_clean['PAY_1'].value_counts()

0                13087
-1                5047
1                 3261
Not available     3021
-2                2476
2                 2378
3                  292
4                   63
5                   23
8                   17
6                   11
7                    9
Name: PAY_1, dtype: int64

In [30]:
valid_pay_1_mask = df_clean['PAY_1'] != 'Not available'
valid_pay_1_mask.head(10)

0    True
1    True
2    True
3    True
4    True
5    True
6    True
7    True
8    True
9    True
Name: PAY_1, dtype: bool

In [31]:
sum(valid_pay_1_mask)

26664

In [32]:
df_clean_2 = df_clean.loc[valid_pay_1_mask, :].copy()
df_clean_2.shape

(26664, 25)

In [34]:
df_clean_2['PAY_1'] = df_clean_2['PAY_1'].astype('int64')
df_clean_2[['PAY_1', 'PAY_2']].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 26664 entries, 0 to 29999
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   PAY_1   26664 non-null  int64
 1   PAY_2   26664 non-null  int64
dtypes: int64(2)
memory usage: 624.9 KB
